drivers/i2c/i2c_sam0.c

```diff
@@ -60,6 +85,50 @@ struct i2c_sam0_dev_data {
 	uint8_t num_msgs;
 };
 
+static int i2c_sam0_recover_bus(const struct device *dev)
+{
+	const struct i2c_sam0_dev_config *cfg = dev->config;
+	pinctrl_soc_pin_t soc_pin_sda, soc_pin_scl;
+
+	/* Save original pin config values */
+	soc_pin_sda = cfg->pcfg->states[PINCTRL_STATE_DEFAULT].pins[SERCOM_I2CM_SDA_PIN_IDX];
+	soc_pin_scl = cfg->pcfg->states[PINCTRL_STATE_DEFAULT].pins[SERCOM_I2CM_SCL_PIN_IDX];
+
+	uint8_t sda_port_idx = SAM_PINMUX_PORT_GET(soc_pin_sda);
+	uint8_t scl_port_idx = SAM_PINMUX_PORT_GET(soc_pin_scl);
+	uint32_t sda_pin = SAM_PINMUX_PIN_GET(soc_pin_sda);
+	uint32_t scl_pin = SAM_PINMUX_PIN_GET(soc_pin_scl);
+	const struct device *sda_dev = sam_port_devs[sda_port_idx];
+	const struct device *scl_dev = sam_port_devs[scl_port_idx];
+
+	gpio_pin_configure(sda_dev, sda_pin, GPIO_INPUT | GPIO_ACTIVE_HIGH);
+	gpio_pin_configure(scl_dev, scl_pin, GPIO_OUTPUT | GPIO_OUTPUT_INIT_HIGH);
+
+	/* Recover bus */
+	for (uint8_t i = 0; i < 9; i++) {
+		if (gpio_pin_get(sda_dev, sda_pin)) {
+			break;
+		}
+		gpio_pin_set(scl_dev, scl_pin, 0);
+		k_usleep(4);
+		gpio_pin_set(scl_dev, scl_pin, 1);
+		k_usleep(4);
+	}
+
+	/* Generate stop condition */
+	gpio_pin_configure(sda_dev, sda_pin, GPIO_OUTPUT | GPIO_OUTPUT_INIT_HIGH);
+	gpio_pin_set(sda_dev, sda_pin, 0);
+	k_usleep(4);
+	gpio_pin_set(sda_dev, sda_pin, 1);
+	k_usleep(4);
+
+	/* Recover pinmux */
+	pinctrl_configure_pins(&soc_pin_sda, 1, PINCTRL_REG_NONE);
+	pinctrl_configure_pins(&soc_pin_scl, 1, PINCTRL_REG_NONE);
+
```

Comment:
Why do you need this? Why not use below?

```
	retval = pinctrl_apply_state(cfg->pcfg, PINCTRL_STATE_DEFAULT);
	if (retval < 0) {
		return retval;
	}
```

Reply:
removed from this PR